### ISSUE TITLE
Set tempo based on recorded loop

### ISSUE THREAD

**Issue by kasbah** [openAVproductions/openAV-Luppp#65]
Luppp's workflow seems to be set on recording to the bpm you set. Coming from SooperLooper I am more used to recording a loop first and then the tempo is set from the initial loop and number of 8ths per loop. 

This is probably quite a big one so I am interested in your thoughts on adding this alternative workflow.

**Comment by Nomys** [comment]
Another vote for this feature +++

**Comment by harryhaaren** [comment]
Hi @Nomys, thanks for your input - I'm not sure if this is going to happen soon I'm afraid. OpenAV is working on some other projects with a higher priority right now (releasing ArtyFX with AVTK UI's, and testing the plugins on the MOD).

I do see the need for a looper that is more musician friendly, and OpenAV does hope to accomodate workflows like yours better in future. Cheers, -Harry

**Comment by Nomys** [comment]
Aww, that's too bad, I will maybe go back to sooperlooper, cause I'm only using live samples and no pre-recorded samples...

Plus it seem to me that the "use as tempo" function is very one sided, cause it's only to use a loop with and unknow tempo. Changing it into a set tempo based on recorded loop, just need to play the unknow tempo loop with an another soft/source, and record it with luppp. More annoying but the function is far more flexible in itself.

Or am I missing something with "use as tempo" ?

**Comment by Nomys** [comment]
Is there any news regarding this ?

It's been pending for ages...

**Comment by georgkrause** [comment]
@Nomys honestly, nobody is working on this. but we would be happy to see a Pull Request with this feature. ;)

**Comment by georgkrause** [comment]
@Nomys Since i got this feature request on LAC a few times and #128 seems to be solved in the near future and this is the oldest feature request we have, i started thinking about this one. 

Before i collect some thoughts on this, i need to say the following: I spend a big amount of time on Luppp development in the last weeks to get rid of this annoying bug. Sadly, this totally blocked my brain and i didnt found time to actually make music. This is no problem at all, but when the bug is fixed some day, I will concentrate again on some musical projects. This means, i cannot tell anything about when this feature will be ready for production.
Before we can actually implement this, I think we need higher precision BPM-settings. We need to set somethink like "120.65". This might be not that hard, but i didn't looked into this.

Anyway, I think its time to start working on this. Yey! This means, i need your help for thinking about the workflow. What we need is a way to tell Luppp to disable the grid on recording for a specific clip. How do you want to do this? What can you imagine? What do you need. Would be happy about any input about.

In the end, this feature wont make it into the next release, but hopefully in the one after the next :)

**Comment by Nomys** [comment]
\o/\o/\o/\o/\o/

I think a sync choice button like in sooperlooper could do the trick for me.
This button allows to change the focus of the grid. Since Luppp is mainly about bpm sync, a sync button will be just a choice between "bpm" or "clip". 
"clip" can have a lot of choices (all clips or just one clip per track made only for this use allowing multiples tempos between tracks), but for starter just have one clip you can record on with the grid disabled, then define the grid at the length of the recording and finally enable the grid for the other loops.

What do you think ?

**Comment by georgkrause** [comment]
> "clip" can have a lot of choices (all clips or just one clip per track made only for this use allowing multiples tempos between tracks)

The current TimeManager does not allow this, and this is not really the topic of this feature request, is it? This would need some pretty deep changes and is already tracked somewhere else. Lets focus on the feature of setting the tempo after recording something in relation to the recorded clip :)

> just have one clip you can record on with the grid disabled, then define the grid at the length of the recording and finally enable the grid for the other loops.

This seems pretty possible. We need a button somewhere to disable the grid. Than record a clip without grid and after this we can use the "Set as tempo"-function for setting the grid again. So we have two steps here: 

1. Make it possible to disable the BPM-Grid.
2. Determine how much beats are in the recorded clip.

After thinking a little about this, I think a simple global "Disable Grid"-Button can cause some problems. It might be better to have a more intergrated workflow. Eg a button which says: Record next clip without grid, determine the BPMs and set tempo.

Maybe this could be also related to #119 somehow. We could set a master clip which can be recorded with or without grid and tempo is aligned to this.

So, i see some problems, i have some ideas, this needs to go into a specific workflow and maybe i also need to do some tests.

@harryhaaren One technical question: Is the event-system fast enough to set the exact start and stop of a clip? So when i press the button on the GUI or the Midi controller, is there any relevant latency until the clip is triggered?

**Comment by Nomys** [comment]
> The current TimeManager does not allow this, and this is not really the topic of this feature request, is it? This would need some pretty deep changes and is already tracked somewhere else. Lets focus on the feature of setting the tempo after recording something in relation to the recorded clip :)

I kind of knew this would be a problem :)

> This seems pretty possible. We need a button somewhere to disable the grid. Than record a clip without grid and after this we can use the "Set as tempo"-function for setting the grid again. So we have two steps here:
> 
> 1. Make it possible to disable the BPM-Grid.
> 2. Determine how much beats are in the recorded clip.
> After thinking a little about this, I think a simple global "Disable Grid"-Button can cause some problems. It might be better to have a more intergrated workflow. Eg a button which says: Record next clip without grid, determine the BPMs and set tempo.

Yes !

I think one problem in this will be the second point, cause you can't set a meaningful tempo with only a sample length... You only can have a relative tempo use by the grid but not visible for the user cause confusing.

**Comment by georgkrause** [comment]
I don't really get what you mean. We have the length of the recorded clip given in samples. We know the BPM is between 60 and 220 (current limits of luppp). So we have some values which are possible. Others are not. The difficult part is to chose. We cannot really wait for user input (or can we?), since it needs to be applied fast. On the other hand: how bad is it to pick the wrong tempo? It actually applied tempo might be twice or half of the intended one. So things would still fit somehow, but your metronome shows 1/8 or 1/2 instead of 1/4. Can we live with that?

I need to investigate how its done on loading an unknown sample... Let's see. Please try this feature, because this is what we already have and using this as a base would make things easier.

**Comment by Nomys** [comment]
Sorry, I wasn't very clear, this : 

> It actually applied tempo might be twice or half of the intended one. So things would still fit somehow, but your metronome shows 1/8 or 1/2 instead of 1/4.

Is what I mean by relative tempo, we can totally live with that, but it's just not as precise as 144.8 bpm for exemple...

> I need to investigate how its done on loading an unknown sample... Let's see. Please try this feature, because this is what we already have and using this as a base would make things easier.

As I recall, You set a tempo and we you load a sample the grid is just applied  to it, if the sample doesn't match the grid, some beats will just be blank.

I will test this week, but I think this is something like this...

**Comment by georgkrause** [comment]
>Is what I mean by relative tempo, we can totally live with that, but it's just not as precise as 144.8 bpm for exemple...

These are two different topics. Of course, before we can start working on this, we need the possibility to set the BPM more precicely. This means things like `145.35` need to be possible. This is the first step.

Applying the grid can than have the issue that you play something on lets say 60 BPM but Luppp assume its 120 BPM. The question is, if this is a problem or if we want a user decision which would force luppp to wait for it.

Let me know how your testing was.

**Comment by Nomys** [comment]
> Applying the grid can than have the issue that you play something on lets say 60 BPM but Luppp assume its 120 BPM. The question is, if this is a problem or if we want a user decision which would force luppp to wait for it.

It is not :)

I didn't have time to test and it will be difficult on the few weeks ahead to find time but I'm still on this...

**Comment by Nomys** [comment]
It seems great !

Is this clip will be a specific clip (in a specific slot) or is it just the first clip recorded after entering the special mode ?

> This way, all already recorded clips will be stretched to fit into the new tempo (like it is now when you change the tempo).

So you have tested ? Clips are stretched when the tempo is changed ? That's pretty odd behaviour for me, but doesn't really change anything for my needs...

**Comment by georgkrause** [comment]
> So you have tested ? Clips are stretched when the tempo is changed ? That's pretty odd behaviour for me, but doesn't really change anything for my needs...

Just run current luppp and change the tempo and you will see the length of the clip is changed to fit again to the new tempo. Why is this odd?

It is not a specific clip, you decide which one it is by recording there.

This has one problem: If you record a new clip this way, the wrong BPM-Choice will make your already recorded clips be played back on the wrong speed.

**Comment by coderkun** [comment]
And in this special mode it is only possible to record one clip and not multiple clips with different lengtth at the same time, @georgkrause?

**Comment by Nomys** [comment]
> This has one problem: If you record a new clip this way, the wrong BPM-Choice will make your already recorded clips be played back on the wrong speed.

It's not really an issue, but rather a feature to me ;o)

> Leaving the special mode after recording allows you to have a workflow like this:

Disable grid.
Record one free, not grid synced clip.
Luppp automatically changes the tempo of the session.
You can instantly continue recording other clips which are in sync to the first recorded clip.

I think it's both easy and flexible this way ! 
Sorry I didn't have time to dig on this this summer, I may have more time this fall :/

**Comment by georgkrause** [comment]
> It's not really an issue, but rather a feature to me ;o)

Its not an feature to play back a clip on twice or half speed as intended. This is just wrong.

**Comment by georgkrause** [comment]
I am currently reworking the clipsync and working on improving the general way how the playback works in several minor step. After everything is working, I plan to implement a finer granularity for BPMs (so we can set something like 139.24 BPM). After we did this, it should be quite easy to align the grid to a in free mode recorded clip. 

This is still quite a long way to go for me alone, so don't expect this to be done soon, but at least there is some progress :)

**Comment by georgkrause** [comment]
@kasbah @magnetophon @Nomys @harryhaaren FYI, I have a first POC: https://peertube.servebeer.com/videos/watch/e92af357-2578-4c3c-9e9b-32ab146aedc3

**Comment by magnetophon** [comment]
@georgkrause Sorry for the late reply!

I watched you video but there was no sound!

Thanks for working on this though!

**Comment by georgkrause** [comment]
@magnetophon Thanks for the interest. This was just a first POC so I only did a first video without any sound. There is a full demo of everything which I planned to release as a fork of Luppp called Loopp: https://peertube.servebeer.com/videos/watch/e52aa296-7aa7-4200-be55-a2d690c8e3f9

But after some long discussions and because of some personal stuff this fork wont be released to the public.

**Comment by magnetophon** [comment]
That looks great!

Too bad you don't want to release it!
Could you tell us more about why you don't?

**Comment by magnetophon** [comment]
BTW, is that a Fusion wristband?  I'm in Lärz as we speak!

**Comment by georgkrause** [comment]
@magnetophon Jeah its from Fusion 2018, needed to skip this years sadly but was there every time from 2014-2018 (except 2017 of course ;).

I did a fork of Luppp called Loopp because I wanted to push the project forward without the limitations of a project which is part of a pack. But after a lot of discussions I think its a bad idea so split the project and its better to get the changes into Luppp. But currently I can't care for it.

**Comment by magnetophon** [comment]
Ah, that sounds much better than what I originally thought you meant! :)

Usually when people say they don't want to publish code *yet*, they want to clean it up first, which is understandable. Is that the case here as well?
Would it be an option to send your changes privately to @harryhaaren so he can get inspired by them?

**Comment by harryhaaren** [comment]
Hey Bart/Georg!

There's various options - but unfortunately I just don't have much time for OpenAV at them moment, and what little time I do have I tend to prioritize to Ctlra lib. I'm open to options as to how we can make Luppp a "community maintained" OpenAV project or something, that gives other developers more/easier control/direction of it - I just don't have the time (and dont' see that changing very soon..).

Thanks for the input though, and if there are suggestions then lets have them! (Perhaps we can open a new ticket if we're expecting a big discussion - just to no derail "set tempo" thread too much! :D )

Cheers, -H

**Comment by georgkrause** [comment]
> Usually when people say they don't want to publish code _yet_, they want to clean it up first, which is understandable. Is that the case here as well?
> Would it be an option to send your changes privately to @harryhaaren so he can get inspired by them?

@magnetophon I already sent the changes for review and we are already in a process of discussing the future of Luppp. I would like to stop discussion here now, to keep the issue clean. For further question, or Fusion small talk, dont hesitate to send an E-Mail or open another issue, please.

**Comment by magnetophon** [comment]
Cool, thanks!
